scripts/manpage-alert.1
.TH MANPAGE-ALERT 1 "Debian Utilities" "DEBIAN" \" -*- nroff -*-
.SH NAME
manpage-alert \- check for binaries without corresponding manpages
.SH SYNOPSIS
\fBmanpage-alert\fR [\fIoptions\fR] [\fIpaths\fR]
.SH DESCRIPTION
\fBmanpage-alert\fR searches the given list of paths for binaries without
corresponding manpages.
.P
If no \fIpaths\fR are specified on the command line, the path list
\fI/bin /sbin /usr/bin /usr/sbin /usr/games\fR will be assumed.
.SH OPTIONS
.TP
.BR \-h\fR, \fB\-\-help
Show a summary of options.
.TP
.BR \-V\fR, \fB\-\-version
Show version and copyright information.
.TP
.BR \-f\fR, \fB\-\-file
Show filenames of missing manpages without any leading text.
.TP
.BR \-p\fR, \fB\-\-package
Show filenames of missing manpages with their package name.
.TP
.BR \-n\fR, \fB\-\-no\-stat
Do not show statistics at the end.
.SH AUTHOR
\fBmanpage-alert\fR was written by Branden Robinson and modified by
Julian Gilbey <jdg@debian.org> and Adam D. Barratt
<debian\-bts@adam\-barratt.org.uk> (who also wrote this manpage) for the
devscripts package.
.P
This manpage and the associated program are licensed under the terms of
the GPL, version 2 or later.


debian/NEWS
devscripts (2.14.0) unstable; urgency=medium

  uscan now looks for upstream's keyring under debian/upstream/ as either
  signing-key.pgp or, for an armored keyring, signing-key.asc.  The previously
  used location, debian/upstream-signing-key.pgp, will be checked as a last
  resort for a transition period.

 -- James McCoy <jamessan@debian.org>  Sat, 25 Jan 2014 10:19:02 -0500

devscripts (2.11.9) unstable; urgency=low

  The default for mk-build-deps --tool option/MKBUILDDEPS_TOOL configuration
  changed from "apt-get" to "apt-get --no-install-recommends".

 -- James McCoy <jamessan@debian.org>  Sun, 17 Jun 2012 22:56:26 +0200

devscripts (2.11.7) unstable; urgency=low

  The heuristic used to determine what action to perform when no options are
  given to debchange now defaults to 'changelog'.  To stay with the previous
  behavior, DEBCHANGE_RELEASE_HEURISTIC=log can be added to the configuration
  file.

  Enabling the changelog heuristic also enables DEBCHANGE_MAINTTRAILER.  This
  can be disabled by adding DEBCHANGE_MAINTTRAILER=no to the configuration
  file.

 -- James McCoy <jamessan@debian.org>  Sun, 29 Apr 2012 18:38:46 -0400

devscripts (2.11.2) unstable; urgency=low

  Support for pgp-style commands has been removed from debsign/debrsign.  As
  such, the DEBSIGN_SIGNLIKE configuration variable has also been removed.

 -- James McCoy <jamessan@debian.org>  Tue, 15 Nov 2011 21:36:40 -0500

devscripts (2.10.54) unstable; urgency=low

    manpage-alert now, as originally intended, prints out the full path to the
    binary which is lacking a manpage instead of simply the binary name.

    Before:
      No manual entry for somebinary

    After:
      No manual entry for /usr/bin/somebinary

 -- James Vega <jamessan@debian.org>  Thu, 06 Aug 2009 18:03:41 -0400

devscripts (2.10.24) unstable; urgency=low

    Starting with devscripts 2.10.21, debchange attempts to
    automatically determine whether an NMU changelog should be
    started, by comparing DEBEMAIL and DEBFULLNAME to the
    package's Maintainer and Uploaders fields.

    This behaviour may now be disabled using the DEBCHANGE_AUTO_NMU
    configuration file variable.

 -- Adam D. Barratt <adam@adam-barratt.org.uk>  Mon, 07 Apr 2008 14:38:08 +0100

devscripts (2.10.20) unstable; urgency=low

    devscripts 2.10.19 modified debcommit's --release option to use the
    uncommitted changes (if any) in the changelog to derive the commit
    message.

    This behaviour is now optional and can be enabled using the -R /
    --release-use-changelog command line option or by setting the
    DEBCOMMIT_RELEASE_USE_CHANGELOG configuration option to "yes".

    The default is to use the previous (i.e. pre-2.10.19) behaviour.

 -- Adam D. Barratt <adam@adam-barratt.org.uk>  Wed, 19 Mar 2008 20:31:20 +0000

devscripts (2.10.19) unstable; urgency=low

    In devscripts 2.10.17, dget was modified to unpack source packages by
    default.  The new -d switch or the DGET_UNPACK configuration file option
    may be used to indicate that the unpacking should not occur.

    As linda has been removed from both unstable and testing, support for
    it has been removed from debuild and cvs-debuild.

 -- Adam D. Barratt <adam@adam-barratt.org.uk>  Thu, 13 Mar 2008 19:33:38 +0000


scripts/manpage-alert.sh
#!/bin/sh -e
#
# Copyright 2005 Branden Robinson
# Changes copyright 2007 by their respective authors
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

usage() {
    cat <<EOF
Usage: manpage-alert [options] [paths]
  Options:
    -h, --help          This usage screen.
    -V, --version       Display the version and copyright information.
    -f, --file          Show filenames of missing manpages
                        without any leading text.
    -p, --package       Show filenames of missing manpages
                        with their package name.
    -n, --no-stat       Do not show statistics at the end.

  This script will locate executables in the given paths with manpage
  outputs for which no manpage is available and its statictics.

  If no paths are specified on the command line, "/bin /sbin /usr/bin
  /usr/sbin /usr/games" will be used by default.
EOF
}

version() {
    cat <<EOF
This is manpage-alert, from the Debian devscripts package, version ###VERSION###.
This code is (C) 2005 by Branden Robinson, all rights reserved.
This program comes with ABSOLUTELY NO WARRANTY.
You are free to redistribute this code under the terms of the
GNU General Public License, version 2 or later.
EOF
}

showpackage() {
    F1="$1"
    P1="$(LANG=C dpkg-query -S "$F1" 2> /dev/null || true )"
    Q1=""; R1=""; Q2=""; R2=""
    if [ -n "$P1" ]; then
        Q1="$(echo "$P1" | grep -v "^diversion by" || true)"
        R1="$(echo "$P1" | sed -ne 's/^diversion by \(.*\) to:.*$/\1/p'): $F1"
    fi
    # symlink may be created by postinst script for alternatives etc.,
    if [ -z "$Q1" ] && [ -L "$F1" ]; then
        F2=$(readlink -f "$F1")
        P2="$(LANG=C dpkg-query -S "$F2" 2> /dev/null || true )"
        if [ -n "$P2" ]; then
            Q2="$(echo "$P2" | grep -v "^diversion by" || true)"
            R2="$(echo "$P2" | sed -ne 's/^diversion by \(.*\) to:.*$/\1/p'): $F2"
        fi
    fi
    if [ -n "$Q1" ]; then
        echo "$Q1"
    elif [ -n "$R1" ]; then
        echo "$R1 (diversion)"
    elif [ -n "$Q2" ]; then
        echo "unknown_package: $F1 -> $Q2"
    elif [ -n "$R2" ]; then
        echo "unknown_package: $F1 -> $R2 (diversion)"
    else
        echo "unknown_package: $F1"
    fi
}

SHOWPACKAGE=DEFAULT
SHOWSTAT=TRUE

while [ -n "$1" ]; do
    case "$1" in
        -h|--help) usage; exit 0;;
        -V|--version) version; exit 0;;
        -p|--package) SHOWPACKAGE=PACKAGE
            shift
            ;;
        -f|--file) SHOWPACKAGE=FILE
            shift
            ;;
        -n|--no-stat) SHOWSTAT=FALSE
            shift
            ;;
        *)  break
            ;;
    esac
done

if [ $# -lt 1 ]; then
    set -- /bin /sbin /usr/bin /usr/sbin /usr/games
fi

NUM_EXECUTABLES=0
NUM_MANPAGES_FOUND=0
NUM_MANPAGES_MISSING=0

for DIR in "$@"; do
    for F in "$DIR"/*; do
        # Skip as it's a symlink to /usr/bin
        if [ "$F" = "/usr/bin/X11" ]; then continue; fi
        NUM_EXECUTABLES=$(( $NUM_EXECUTABLES + 1 ))

        if OUT=$(man -w -S 1:8:6 "${F##*/}" 2>&1 >/dev/null); then
            NUM_MANPAGES_FOUND=$(( $NUM_MANPAGES_FOUND + 1 ))
        else
            if [ $SHOWPACKAGE = "PACKAGE" ]; then 
                # echo "<packagename>: <filename>"
                showpackage "$F"
            elif [ $SHOWPACKAGE = "FILE" ]; then
                # echo "<filename>"
                echo "$F"
            else
                # echo "No manual entry for <filename>"
                echo "$OUT" | perl -ne "next if /^.*'man 7 undocumented'.*$/;" \
                  -e "s,(\W)\Q${F##*/}\E(?:\b|$),\1$F,; s,//,/,; print;"
            fi
            NUM_MANPAGES_MISSING=$(( $NUM_MANPAGES_MISSING + 1 ))
        fi
    done
done

if [ $SHOWSTAT = "TRUE" ]; then 
echo
printf "Of %d commands, found manpages for %d (%d missing).\n" \
    $NUM_EXECUTABLES \
    $NUM_MANPAGES_FOUND \
    $NUM_MANPAGES_MISSING
fi

# vim:set ai et sw=4 ts=4 tw=80:
